### PR TITLE
updating CTS units and capacitance/slew intervals.

### PR DESCRIPTION
**Pull request by Mahfouz-z**
Disambiguating cts units and removing unecssary extra units dimensions. Editing capacitance and slew steps.

### CODE REVIEW

**Comment by Mahfouz-z** [comment]
@maliberty the tests are still not passing as I did not update the .ok files until we make sure the logic is correct.

File: src/cts/src/HTreeBuilder.cpp [hunk 1]
```diff
@@ -157,7 +157,7 @@ void HTreeBuilder::preSinkClustering(
 
 void HTreeBuilder::initSinkRegion()
 {
-  unsigned wireSegmentUnitInMicron = techChar_->getLengthUnit();
+  unsigned wireSegmentUnitInMicron = techChar_->getLengthUnitMicron();
```
Comment (maliberty) [inline src/cts/src/HTreeBuilder.cpp:160]: Yikes - microns in unsigned will truncate.

**Comment by Mahfouz-z** [comment]
@maliberty In this commit, I tried to eliminate the use of Microns in TechChar and HTreeBuilder. I also added your note that the user should input the count of capacitance and slew intervals instead of their deltas.

File: src/cts/src/CtsOptions.h [hunk 2]
```diff
@@ -67,13 +67,13 @@ class CtsOptions
     bufferList_ = buffers;
   }
   std::vector<std::string> getBufferList() const { return bufferList_; }
-  void setDbUnits(int units) { dbUnits_ = units; }
-  int getDbUnits() const { return dbUnits_; }
-  void setWireSegmentUnit(unsigned wireSegmentUnit)
+  void setDbUnits(float units) { dbUnits_ = units; }
```
Comment (maliberty) [inline src/cts/src/CtsOptions.h:70]: Why is this a float now?

File: src/cts/src/HTreeBuilder.cpp [hunk 3]
```diff
@@ -367,7 +375,7 @@ void HTreeBuilder::computeLevelTopology(unsigned level,
     remainingLength = previousLevel.getRemainingLength();
   }
 
-  const unsigned SLEW_THRESHOLD = options_->getMaxSlew();
+  const unsigned SLEW_THRESHOLD = techChar_->getMaxSlew(); // The normalized value of the slew of the characterization buffer (max slew)
```
Comment (maliberty) [inline src/cts/src/HTreeBuilder.cpp:378]: Why is this stored in two different places?  Can we consolidate this to just one?

File: src/cts/src/TechChar.cpp [hunk 4]
```diff
@@ -551,28 +536,32 @@ void TechChar::initCharacterization()
 
   // Defines the different wirelengths to test and the characterization unit.
   unsigned wirelengthIterations = options_->getCharWirelengthIterations();
-  unsigned maxWirelength = (charBuf_->getHeight() * 10)
-                           * wirelengthIterations;  // Hard-coded limit
-  if (options_->getWireSegmentUnit() == 0) {
-    unsigned charaunit = charBuf_->getHeight() * 10;
-    options_->setWireSegmentUnit(charaunit);
+  if (options_->getWireSegmentUnitDbu() == 0) {
+    unsigned charaunitDbu = charBuf_->getHeight() * 10;
```
Comment (maliberty) [inline src/cts/src/TechChar.cpp:540]: charaunitDbu is an odd name - maybe wireSegmentUnitDbu

File: src/cts/src/TechChar.cpp [hunk 5]
```diff
@@ -551,28 +536,32 @@ void TechChar::initCharacterization()
...
+    options_->setWireSegmentUnitDbu(charaunitDbu);
   } else {
-    // Updates the units to DBU.
-    unsigned segmentDistance = options_->getWireSegmentUnit();
-    options_->setWireSegmentUnit(segmentDistance * dbUnitsPerMicron);
+    // user has passed the segment unit in Micron, so we need to convert
+    // it to DBU to keep correctness and uniformity.
+    options_->setWireSegmentUnitDbu(options_->getWireSegmentUnitDbu()
+                                    * dbUnitsPerMicron);
```
Comment (maliberty) [inline src/cts/src/TechChar.cpp:546]: The conversion from microns to dbu should be done on the way in from tcl rather the delayed to here.  I want to remove microns as much as possible from the core code.

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/cts/src/CtsOptions.h (hunk 2 above)
Comment (Mahfouz-z) [reply src/cts/src/CtsOptions.h:70]: TechChar.cpp line 447 confused me that the value is actually a float but yes, it should be an int.

File: src/cts/src/HTreeBuilder.cpp (hunk 3 above)
Comment (Mahfouz-z) [reply src/cts/src/HTreeBuilder.cpp:378]: Yeah makes since to make a single source for that, actually, PostCtsOpt was still using the max from options_ too.

File: src/cts/src/CtsOptions.h (hunk 2 above)
Comment (Mahfouz-z) [reply src/cts/src/CtsOptions.h:70]: Sorry, I confused it.

File: src/cts/src/CtsOptions.h [hunk 6]
```diff
@@ -209,20 +207,19 @@ class CtsOptions
...
   bool simpleCts_ = false;
   bool sinkClusteringEnable_ = true;
   bool sinkClusteringUseMaxCap_ = true;
   bool simpleSegmentsEnable_ = false;
   bool vertexBuffersEnable_ = false;
   bool gui_debug_ = false;
-  double vertexBufDistance_ = 240;
-  double bufDistance_ = 100;
+  double vertexBufDistance_ = 0; // in DBU 
+  double bufDistance_ = 0; // in DBU
```
Comment (rovinski) [inline src/cts/src/CtsOptions.h:219]: Comment is DBU but type is double? Also why change value to 0?

File: src/cts/src/HTreeBuilder.cpp [hunk 7]
```diff
@@ -33,14 +33,15 @@
 //
 ///////////////////////////////////////////////////////////////////////////////
 
+#include "HTreeBuilder.h"
```
Comment (rovinski) [inline src/cts/src/HTreeBuilder.cpp:36]: Why is this moved?

File: src/cts/src/TechChar.cpp [hunk 8]
```diff
@@ -33,13 +33,14 @@
 //
 ///////////////////////////////////////////////////////////////////////////////
 
+#include "TechChar.h"
```
Comment (rovinski) [inline src/cts/src/TechChar.cpp:36]: Same

File: src/cts/src/TechChar.cpp [hunk 9]
```diff
@@ -451,18 +444,10 @@ void TechChar::initCharacterization()
 {
   odb::dbChip* chip = db_->getChip();
   odb::dbBlock* block = chip->getBlock();
-  float dbUnitsPerMicron = block->getDbUnitsPerMicron();
-
+  int dbUnitsPerMicron = block->getDbUnitsPerMicron();
```
Comment (rovinski) [inline src/cts/src/TechChar.cpp:447]: Need to be careful because `getClockLayerResCap()` now takes in `int`, not `float`

File: src/cts/src/TritonCTS.i [hunk 10]
```diff
@@ -141,14 +144,14 @@ void
 set_distance_between_buffers(double distance)
 {
   getTritonCts()->getParms()->setSimpleSegmentsEnabled(true);
-  getTritonCts()->getParms()->setBufferDistance(distance);
+  getTritonCts()->getParms()->setBufferDistance(distance * getDb()->getChip()->getBlock()->getDbUnitsPerMicron());
```
Comment (rovinski) [inline src/cts/src/TritonCTS.i:147]: When converting from um to DBU I would recommend using `std::round()` because the implicit cast from float to int will truncate.

**Review by rovinski — COMMENTED** [review]
Needs another clang format I believe

File: src/cts/src/CtsOptions.h (hunk 6 above)
Comment (Mahfouz-z) [reply src/cts/src/CtsOptions.h:219]: These two entries used to be in Microns; therefore, they were double, will convert them to ints. I changed the value to 0 because it was set to a value that could be considered as default; however, it was not really used at all as these values are only used when a boolean is activated that depends on these values being passed in the tcl.

File: src/cts/src/HTreeBuilder.cpp (hunk 7 above)
Comment (Mahfouz-z) [reply src/cts/src/HTreeBuilder.cpp:36]: It was moved after running clang-format.

File: src/cts/src/TechChar.cpp (hunk 8 above)
Comment (Mahfouz-z) [reply src/cts/src/TechChar.cpp:36]: same here

File: src/cts/src/TechChar.cpp (hunk 9 above)
Comment (Mahfouz-z) [reply src/cts/src/TechChar.cpp:447]: static cast to float will do the trick, right?

File: src/cts/src/TritonCTS.i (hunk 10 above)
Comment (Mahfouz-z) [reply src/cts/src/TritonCTS.i:147]: Done.

File: src/cts/src/HTreeBuilder.cpp (hunk 7 above)
Comment (rovinski) [reply src/cts/src/HTreeBuilder.cpp:36]: clang-format moved it from the correct order to out-of-order? That doesn't seem right...

**Comment by maliberty** [comment]
obsolete
